Refactor to spec

ethanfrey opened the pull request:
This implements all parts of the README except for the download functionality (will be a separate PR). Actually download code is writen (copied from original design), but not tested.

Includes unit test coverage for all elements, including argument pass through, responding to log lines, exiting properly on normal exit (no upgrade), and switching directories on the proper triggers, along with many sanity checks.

I would like to get a review on this and merge it in, before proceeding in two more directions:

* Thorough tests on the download functionality
* Manual integration tests with real blockchain daemons

ethanfrey committed:
Minor fix to log regexp

ethanfrey committed:
Wrote code for parsing and using config from env

ethanfrey committed:
Updated go-getter to v1.4.0

ethanfrey committed:
Add scanner file to abstract out parsing

ethanfrey committed:
Start writing minimal main with error checks

ethanfrey committed:
Wrote basic process listening code

ethanfrey committed:
Code upgrade and auto-download process

ethanfrey committed:
Add first tests

ethanfrey committed:
Test config validation

ethanfrey committed:
Add testdata, test EnsureBinary checks

ethanfrey committed:
Add test for scanner, fix close response

ethanfrey committed:
Fix regexp to handle spaces better, chunks

ethanfrey committed:
Test and fix SetCurrentUpgrade

ethanfrey committed:
Allow updating CurrentUpgrade multiple times

ethanfrey committed:
Cleanup

ethanfrey committed:
Modify CurrentBin() to resolve link for easier test/debugging

ethanfrey committed:
Test DoUpgrade without download

ethanfrey committed:
Test Cleanup

ethanfrey committed:
Cleanup LaunchProcess, add into to dummy bins

ethanfrey committed:
WIP: got LaunchProcess to terminate, but lost stdout

ethanfrey committed:
Cleanup unneeded complexity, simpler Pipe->Tee->Scanner arch

aaronc commented on args_test.go:
```diff
@@ -0,0 +1,134 @@
+	"github.com/stretchr/testify/assert"
+)
+
+// Note all tests use the / path separator for simplicity and will fail on windows.
```
Could we avoid the technical debt now and just use https://golang.org/pkg/path/filepath/#FromSlash ?

ethanfrey replied:
Ah, niceone. I did not know that and just wanted to use strings to test, instead of tons of filpath.Join(). I will update.

aaronc commented on process.go:
```diff
@@ -0,0 +1,111 @@
+	// wait for everyone to finish....
+	wg.Wait()
+	return info, err
+}
```
Could you add just a few more comments in this function explaining why you chose to use a mutex and wait group to accomplish this functionality? That will make it clearer for me as well as future people working on this code.

ethanfrey replied:
3 go routines all touching the same variables? Basically I want a return value from 3 threads... 

I will document a bit more and maybe we can simplify... if we just wait for the command to exit, and assume it will not print out the upgrade log line and exit within a microsecond or two later, maybe I can just run that sync and block on the command...

This was the trickiest part for me, as there are three conditions to end, and all need to be checked in different threads.

ethanfrey replied:
Addressed this in 24e2ace

aaronc commented on upgrade.go:
```diff
@@ -0,0 +1,128 @@
+
+// DoUpgrade will be called after the log message has been parsed and the process has terminated.
+// We can now make any changes to the underlying directory without interferance and leave it
+// in a state the will make a proper restart
```
Grammar issue: `in a state the will make a proper restart`

aaronc commented on args.go:
```diff
@@ -0,0 +1,111 @@
+type Config struct {
+	Home     string
+	Name     string
+	Download bool
```
Maybe rename this `AllowDownloadBinaries` or something that is a little clearer?

aaronc commented on args.go:
```diff
@@ -0,0 +1,111 @@
+		Home: os.Getenv("DAEMON_HOME"),
+		Name: os.Getenv("DAEMON_NAME"),
+	}
+	if os.Getenv("DAEMON_DOWNLOAD") == "on" {
```
Ditto for the env var, how about `DAEMON_ALLOW_DOWNLOAD_BINARIES`?

ethanfrey replied:
Sure, I took this from the README.
I will update names in both places.

aaronc commented on upgrade_test.go:
```diff
@@ -0,0 +1,89 @@
+	}
+}
+
+// TODO: test with download (and test all download functions)
```
Btw with go-getter you could test "downloads" with local files actually

ethanfrey replied:
Nice. I will try.
I was going to add some files to the repo and attempt to get them from github rawcontent.
But local files is a tighter test loop

aaronc commented on scanner_test.go:
```diff
@@ -0,0 +1,88 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func ExampleScanner() {
```
Is this used at all? Why keep here?

ethanfrey replied:
I was having issues with pipes and trying to debug it, trying out different setup (earlier pipe config caused the scanner to miss the prefix of every line for some odd reason - I then simplified it witha  few standard mods from io).

I can remove it.

aaronc reviewed (changes requested):
This is great. Really well structured and I see that you've found creative ways to test otherwise hard to test things.

I requested a few minor changes and had a few questions, but other than that mostly looks good to go.

The one behavioral thing I'm still questioning is whether or not the upgrade manager should just simply launch new sub-processes for upgrades rather than exiting and waiting for systemd to restart the upgrade manager. I think a good compromise solution would be to add an additional config flag that controls whether it exists or keeps running on upgrades (I don't really care which is the default). Seems like it would be a relatively simple change and allow for both operational methods - in particular just testing out a blockchain running straight in my terminal which I often do for testing. What do you think of that approach?

ethanfrey commented:
> The one behavioral thing I'm still questioning is whether or not the upgrade manager should just simply launch new sub-processes for upgrades rather than exiting and waiting for systemd to restart the upgrade manager. I think a good compromise solution would be to add an additional config flag that controls whether it exists or keeps running on upgrades (I don't really care which is the default). Seems like it would be a relatively simple change and allow for both operational methods - in particular just testing out a blockchain running straight in my terminal which I often do for testing. What do you think of that approach?

I thought of this as well - and we can do something if LaunchProcess returns nil - meaning everything is okay - and then just run it again in a loop. Not hard to code.

My worry was all those stdin/stderr writers and pipes and i dunna about memory usage, so figured it is safest just to restart and get fresh connections everywhere.

But the option of another flag is a good idea. `DAEMON_RESTART=on` will cause auto restart in Run if LaunchProcess returns nil. Sounds good. If the binary crashes or the upgrade fails, it will exit. But if we just have successful upgrades, it will keep running.

I would like to start with this default off, like download, but maybe we can later make them default on and have to set them `off` explicitly. Once we have tested them more thoroughly - default settings are simplest for now.

ethanfrey committed:
Incorporated many PR comments

ethanfrey committed:
Simplified and commented WaitForUpgradeOrExit

ethanfrey committed:
Add RestartAfterUpgrade option

ethanfrey commented:
I addressed all pr comments except those on download.
I will add download functionality in a separate PR.

If you are okay with these changes, please approve and merge. Future PRs can be smaller and targeted. I prefer not to load more into this one.
